packages/rxsocketio/src/creators/io.js

```diff
@@ -16,62 +17,86 @@ const defaultOptions = {
   upgrade: true,
 };
 
-// how to handle Auth? ideally, like this: {query: {token}}, which does this: http://localhost?token=abc
-// how to handle WSS?
+// maps socket.io events to RxJS events
+const eventHandlerMap = {
+  connect: () => actions.connect(),
+  connect_error: error => actions.connectError({error}),
+  connect_timeout: timeout => actions.connectTimeout({timeout}),
+  disconnect: reason => actions.disconnect({reason}),
+  reconnect: attemptNumber => actions.reconnect({attemptNumber}),
+  reconnecting: attemptNumber => actions.reconnecting({attemptNumber}),
+  reconnect_error: error => actions.reconnectError({error}),
+  reconnect_failed: attemptNumber => actions.reconnectFailed({attemptNumber}),
+  ping: () => actions.ping(),
+  pong: latency => actions.pong({latency}),
+};
+
+// attach event listeners to Socket.io client
+function socketEventReducer(client, obs) {
+  return (acc, [eventName, handler]) => (
+    client.on(eventName, (...args) => obs.next(handler(...args)))
+  );
+}
+
+// emit messages to a topic
+function createTopicEmitter(obs, _actions = actions) {
+  return topic => message => obs.next(_actions.newMessage({topic, message}));
+}
+
+// handle all messages to a topic
+function handleTopic(client, obs, emitMessage) {
+  return topic => client.on(topic, emitMessage(topic));
+}
+
+// convert Socket.io client to RxJS Observable
+function createObservableFromSocket(
+  client,
+  topics,
+  _actions = actions,
+  _eventHandlerMap = eventHandlerMap,
+) {
+  const topicBlacklist = Object.keys(_eventHandlerMap);
+  if (topicBlacklist.find(key => topics.indexOf(key) > -1)) {
+    return throwError(new Error(
+      `These topics are reserved and cannot be subscribed to: ${topicBlacklist}`
+    ));
+  }
+  return new Observable(obs => {
+    obs.next(_actions.createClient({client}));
+    // map Socket.io events to Observable events
+    toPairs(_eventHandlerMap).reduce(socketEventReducer(client, obs), undefined);
```

Comment:
I like how this part turned out...  since `socketEventReducer` returns a reducer function, it's very concise to call it!

Reply:
This is clearer to me as well, I like it.